### ISSUE TITLE
Define the limits and features of message searching

### ISSUE THREAD

**Issue by webinista** [department-of-veterans-affairs/messaging-fe#40]
- How fuzzy or exact can we be?
- Tolerance for misspellings?
- Full vs partial matches?

Blocks / affects #14, #4

**Comment by ayaleloehr** [comment]
If "exact match" is not selected, we will treat query terms as being OR'd together. If "exact match" is selected, we will treat query terms as being AND'd.  

@angel I'm thinking maybe "exact match" should be next to and apply to the Subject and/or From fields individually, rather than being a single checkbox for the whole search?  (single checkbox feels off since Date Range won't actually change whether or not you check that box I'm assuming?)

UPDATED COMMENT: I think doing AND'ing for when "exact match" is selected is wrong actually. if a query is "test query" and a subject line is "i want to test a query", if we AND the query string (i.e. ask if it matches for both "test" and "query"), this does match. However "exact match" to me seems to specify that the subject should literally contain "test query" and thus "i want to test a query" does not match it.  So, I would like to update the top paragraph here to say "If 'exact match' is selected, we will treat the query as a single string and see if this entire string is contained within the name or message subject.  Leaving the original comment to show my thought process evolving. :)

**Comment by ayaleloehr** [comment]
Also, when a user does a non-advanced search, my assumption is that the API should do an OR on each term and search over both from and subject. 

We will not have a tolerance for misspellings in the MVP; just simple text matching.

**Comment by ahelkit** [comment]
@ayaleloehr the single string is indeed what we want for exact matches. 

As for the exact match, you are also correct. Exact match doesn't apply to dates. Here is an updated design with exact match checkboxes at an individual level. 

**Desktop:**
![advanced search](https://cloud.githubusercontent.com/assets/103496/18457526/759b57f0-790e-11e6-9a24-2e7903e8ac30.png)

**Mobile:**
![sm mobile - adv search expanded](https://cloud.githubusercontent.com/assets/103496/18457740/f796e4da-790f-11e6-9005-b3457e7d050b.png)

**Comment by ayaleloehr** [comment]
@webinista do you think this is well answered?  What open questions still exist?

**Comment by webinista** [comment]
I think these are well-answered by design. Now I guess my questions are around implementation. Am I correct in my understanding that we're going to be searching and filtering messages on the client side?

**Comment by ayaleloehr** [comment]
We will have query parameters to allow the sorting/searching/filtering via the API, so you shouldn't have to do that client side. Pinging @mphprogrammer in slack to get documentation on this out ASAP.

**Comment by ayaleloehr** [comment]
@mphprogrammer please update this issue when you have documented how API clients should access sorting and searching via query parameters.

**Comment by saneshark** [comment]
I'm a little confused on these mocks are going to work. There is still a lot of confusion to me on the following points:
- Exact Match is really confusing for a number of reasons: a) is it case-sensitive vs case-insensitive, b) does it mean that we're doing AND vs OR on matches within the string? All of these things are added complexity considering we have to do this in memory. At most we can do case-sensitive vs case-insensitive and a SANITIZED regex style match for whatever string is provided.
- What is the search at the top for? We can't search for message body at all, we're only limited to from and subject and date range. I think the search bar at the top should possibly go away for advanced search options as it suggests that we can search within the message body.

If multiple inputs are provided we need to be consistent. So we're proposing doing an AND when both subject, from, and date ranges are provided. Doing an either / or is unnecessarily confusing. A user can provide one or the other, and if their search returns too many results they will add additional search criteria to limit it to a smaller subset. But by doing an OR there is no way to limit it really and the results will be very unclear.

**Comment by webinista** [comment]
Any updates here, @mphprogrammer @saneshark?

**Comment by mphprogrammer** [comment]
Filtering on a single field is complete and documented. However, sorting on 2+ fields, while working, has not been documented, and the way we handle that may change. Sorting of Messages, Prescriptions, and Trackings is completed and documented. Sorting of Folder names will be removed in the next PR as it interferes with the preferred order of presentation (per @U-DON).

https://github.com/department-of-veterans-affairs/vets-api/blob/master/docs/api/sm-api.md

**Comment by webinista** [comment]
Just echoing [Kam's comment](https://github.com/department-of-veterans-affairs/messaging-fe/issues/40#issuecomment-250549098) above. 

Since we can't search the message body, it makes no sense to have the basic search box. I can build it, but I think we should drop it.

**Comment by ayaleloehr** [comment]
@angel is out this week.  Tagging @gnakm and @emilyville for thoughts.  I know neither of you have a ton of context, but the issue should hopefully be laid out pretty clearly above.

TL;DR is that the problem is that having the big search box makes it seem like users can search the body of messages, which the API does not allow. Then also when we do have advanced search opened up, the big search bar isn't actually functional and is weird to have 2 search buttons. 

So I think the question is would it make more sense to remove the big search bar (and button with it) completely and only have the advanced fields when someone selects that they want to search, perhaps changing that wording from "advanced search" to just "search"?

![screen shot 2016-10-25 at 4 08 01 pm](https://cloud.githubusercontent.com/assets/4719765/19702370/3e6cf514-9acd-11e6-9e9b-8349e40b5d47.png)

**Comment by gnakm** [comment]
What about disabling the search box when the advanced search options are open?
https://marvelapp.com/11ddc8f/screen/16232723

I think taking the search box completely out will create a weird transition and the purpose of the advance search link is to create an expanding effect. We could definitely ask users this Thurs/Fri. If it's frustrating or confusing, we can take it out. My hypothesis is that there is enough information for users to know that they cannot use the bigger search box at the moment but can be used other times.

**Comment by gnakm** [comment]
I think we should take out big search when advanced search is open. https://marvelapp.com/46edi9j/screen/16232523(click advanced search)

My initial suggestion was going to be disabling it but that will cause 508 problems.
